sympy/integrals/integrals.py

```diff
@@ -397,6 +397,12 @@ def doit(self, **hints):
         if not hints.get('integrals', True):
             return self
 
+        # fixes issue 20693
+        from sympy import Function, Symbol
+        if isinstance(self.args[1][0], Function):
+                t = Symbol('t')
+                return self.subs(self.args[1][0], t).doit(**hints).subs(t, self.args[1][0])
```

Comment:
So, basically this is just a way to substitute the instances of a function with a new variable to `t`. In other words substitution technique. This may not work if there are different functions in the same expression. See, https://www.wolframalpha.com/input/?i=integrate+sin%28f%28x%29+%2B+g%28x%29%29d%28f%28x%29%29%2Fdx as well.
```python
In [1]: f = Function('f')

In [2]: g = Function('g')

In [3]: Integral(sin(f(x) + g(x)), f(x))
Out[3]: 
⌠                         
⎮ sin(f(x) + g(x)) d(f(x))
⌡                         

In [4]: Integral(sin(f(x) + g(x)), f(x)).doit()
Out[4]: -cos(f(x) + g(x))

In [6]: Integral(sin(f(x) + g(x))*f(x).diff(x), x)
Out[6]: 
⌠                             
⎮                  d          
⎮ sin(f(x) + g(x))⋅──(f(x)) dx
⎮                  dx         
⌡                             

In [7]: _.doit()
Out[7]: 
⌠                             
⎮                  d          
⎮ sin(f(x) + g(x))⋅──(f(x)) dx
⎮                  dx         
⌡                    
```
The problem is that when `f(x)` got substituted with `t`, the variable with respect to which integration was supposed to happen changed from `f(x)` to `t` making `g(x)` a constant which shouldn't have been the case. The integral should have left un-evaluated due to abstract nature of `f(x)` and `g(x)`.

Reply:
So, we can substitute `x` with the inverse of `f(t)` i.e. `F(t)`, this way, g(x) will be `g(F(t))` and now, it won't be treated as a constant and we will be able to differentiate w,r,t to `t`, which solves both the problems. 
Is there already some method to find the inverse of a function or we need to implement it?

Reply:
I am suggesting something like:
```
x=list(self.args[1][0].free_symbols)[0]
F= Function('F')
F=inverse(f)
return self.subs(x, F(t)).doit(**hints).subs(t, self.args[1][0])
```
`x` is the variable i.e. `x` in `f(x)` in the integrand, `f(F(t))` will be `t`.
So, in `Integral(sin(f(x) + g(x)), f(x)).doit()` will become `Integral(sin(t + g(F(t))), t).doit()` and later `t` will be substituted by `f(x)`.

Reply:
Well, the general problem is handling `Integral(F(x1, x2, x3, ..., xn), f(x1, x2, x3, ..., xm))` which is equivalent to, `Integral(F(x1, x2, x3, ..., xn)*differential(f(x1, x2, x3, ..., xm)), (x1, x2, x3, ..., xm))`. 
Anyways, this is still a workaround. I would suggest the same as @oscarbenjamin i.e., finding the root cause and fixing it.